crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -506,61 +504,57 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
-
-            let new_content = TimelineItemContent::Message(Message {
-                msgtype,
-                in_reply_to: msg.in_reply_to.clone(),
-                thread_root: msg.thread_root.clone(),
-                edited: true,
-                mentions: replacement.new_content.mentions,
-            });
-
-            let edit_json = match &this.ctx.flow {
-                Flow::Local { .. } => None,
-                Flow::Remote { raw_event, .. } => Some(raw_event.clone()),
-            };
+        if self.ctx.sender != item.sender() {
+            info!(
+                original_sender = ?item.sender(), edit_sender = ?self.ctx.sender,
+                "Edit event applies to another user's timeline item, discarding"
+            );
+            return;
+        }
 
-            trace!("Applying edit");
+        let TimelineItemContent::Message(msg) = item.content() else {
+            info!(
+                "Edit of message event applies to {:?}, discarding",
+                item.content().debug_string(),
+            );
+            return;
+        };
 
-            if let EventTimelineItemKind::Remote(remote_event) = &event_item.kind {
-                let new_encryption_info = match &this.ctx.flow {
-                    Flow::Local { .. } => None,
-                    Flow::Remote { encryption_info, .. } => encryption_info.clone(),
-                };
+        let mut msgtype = replacement.new_content.msgtype;
+        // Edit's content is never supposed to contain the reply fallback.
+        msgtype.sanitize(DEFAULT_SANITIZER_MODE, RemoveReplyFallback::No);
 
-                Some(event_item.with_content(new_content, edit_json).with_kind(
-                    EventTimelineItemKind::Remote(
-                        remote_event.with_encryption_info(new_encryption_info),
-                    ),
-                ))
-            } else {
-                Some(event_item.with_content(new_content, edit_json))
-            }
+        let new_content = TimelineItemContent::Message(Message {
+            msgtype,
+            in_reply_to: msg.in_reply_to.clone(),
+            thread_root: msg.thread_root.clone(),
+            edited: true,
+            mentions: replacement.new_content.mentions,
         });
 
-        if !found {
-            debug!("Timeline item not found, discarding edit");
+        let edit_json = match &self.ctx.flow {
+            Flow::Local { .. } => None,
+            Flow::Remote { raw_event, .. } => Some(raw_event.clone()),
+        };
+
+        let mut new_item = item.with_content(new_content, edit_json);
+
+        if let EventTimelineItemKind::Remote(remote_event) = &item.kind {
+            if let Flow::Remote { encryption_info, .. } = &self.ctx.flow {
+                new_item = new_item.with_kind(EventTimelineItemKind::Remote(
+                    remote_event.with_encryption_info(encryption_info.clone()),
+                ));
+            }
```

Comment:
A small note: this is slightly changing behavior, because I thought the previous behavior was buggy.

Previous behavior did this: if the item comes from a local echo, use an `encryption_info` set to `None`, and replace it in the original timeline item.

With the new behavior, we override the `encryption_info` only if the new information comes from a remote echo.

Incidentally, @pixlwave opened an issue which *I think* describes that, today: #3882. Happy to move the change to a separate PR if it's preferred; since it was only a 3 lines change, I deemed it not useful. But it's worth highlighting for the review here, since this is the only functional change of this overall PR.

Reply:
Yep, I can confirm that this fixes the issue 👍

Reply:
I'm fine keeping the change here.